### PR TITLE
[testutils] Prevent deadlocks in Latch and Await

### PR DESCRIPTION
**Pull request by jamesonwilliams**
`Await.java`: if `Await.result(...)` is called from thread A, and its method
fires callbacks on that same thread A, they may never be invoked. This
is because the `CountDownLatch` is sleeping the thread A. To address this
situation, ensure that the `ResultErrorEmitter` function always run on a
different (new) thread.

`Latch.java`: `InterruptedException`was being gobbled. But, this needs to be
raised to crash the thread. Otherwise, the thread can hang.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: testutils/src/main/java/com/amplifyframework/testutils/Await.java [hunk 1]
```diff
@@ -134,28 +134,37 @@ private Await() {}
...
+                        latch.countDown();
+                    }
+                );
+            } catch (Throwable unexpectedFailure) {
+                unexpectedErrorContainer.set(unexpectedFailure);
                 latch.countDown();
             }
-        );
+        });
+        thread.setDaemon(true);
```
Comment (raphkim) [inline testutils/src/main/java/com/amplifyframework/testutils/Await.java:155]: is there any reason why you want these threads to persist? I feel like all threads should be stopped if the test suite that started it crashes due to a certain error that is unrelated to this thread.

File: testutils/src/main/java/com/amplifyframework/testutils/Await.java [hunk 2]
```diff
@@ -134,28 +134,37 @@ private Await() {}
...
+                        latch.countDown();
+                    }, error -> {
+                        errorContainer.set(error);
+                        latch.countDown();
+                    }
+                );
+            } catch (Throwable unexpectedFailure) {
+                unexpectedErrorContainer.set(unexpectedFailure);
                 latch.countDown();
             }
```
Comment (raphkim) [inline testutils/src/main/java/com/amplifyframework/testutils/Await.java:153]: would anything ever get thrown directly to this thread?

File: testutils/src/main/java/com/amplifyframework/testutils/Await.java (hunk 1 above)
Comment (jamesonwilliams) [reply testutils/src/main/java/com/amplifyframework/testutils/Await.java:155]: Ahhh!!! Good catch. I set out to make this say _`false`_, (`true` is the default anyway.) But, this does _not_ say `false`, does it.

File: testutils/src/main/java/com/amplifyframework/testutils/Await.java (hunk 2 above)
Comment (jamesonwilliams) [reply testutils/src/main/java/com/amplifyframework/testutils/Await.java:153]: ___Not___ if it is working properly. Before, if something was broken, it would throw on the calling thread, here. But since I've moved the call onto a new thread, I don't want the exception to just kill the thread, while the caller is still alive waiting for a response. I want it to kill the _calling_ thread, too.

But yea - this is a safeguard. Software that is functioning as we intent it do would not trigger this condition. (But, this is after-all a _testutils_ component, so we can't expect things passed into this method will always be functioning correctly (they may throw instead of calling error consumer, accidentally.))

File: testutils/src/main/java/com/amplifyframework/testutils/Await.java (hunk 1 above)
Comment (raphkim) [reply testutils/src/main/java/com/amplifyframework/testutils/Await.java:155]: ah okay. that sounds like a good reason.

File: testutils/src/main/java/com/amplifyframework/testutils/Await.java (hunk 2 above)
Comment (raphkim) [reply testutils/src/main/java/com/amplifyframework/testutils/Await.java:153]: I guess it's always better safe than sorry. I just couldn't think of a scenario where anything would ever be thrown directly here now that you created a new well-contained thread.
